File: mpi/solution/pi_mpi.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

int main (int argc, char* argv[]) {
  long long int i, n=100000000;
  double start_time, end_time;
  double x, pi;
  double sum = 0.0, result = 0.0;
  double step = 1.0/(double) n; 
  int rank, size;
	
  /* Initialize MPI */
  MPI_Init(&argc, &argv);
	
  /* Get my process ID number */
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
	
  /* Get my process number */
  MPI_Comm_size (MPI_COMM_WORLD, &size);
	
  /* Start timer only on main */
  if ( rank == 0)  
    start_time = MPI_Wtime();
 
  /* Broadcast value of n to all processes */
  MPI_Bcast(&n,1,MPI_INTEGER,0,MPI_COMM_WORLD);

  /* Divide work such that each process gets n/n_p elements to calculate */ 
  for (i = rank*n/size ; i < (rank + 1)*n/size ; i++ ) {
    x = (i+0.5)*step;
    result +=  4.0/(1.0+x*x);
  }
	
  /* collect all partial results with a reduction operation */
  MPI_Reduce(&result, &sum, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);

  /* Synchronize all processes */
  MPI_Barrier(MPI_COMM_WORLD);

  /* Calculate value of pi and stop timer if main process */	
  if ( rank == 0) {
    end_time = MPI_Wtime();
    pi = step * sum;
    printf("pi = %17.15f\n",pi);
    printf("time to compute = %g seconds\n", (double)(end_time - start_time));
  }
	
  /* Finalize MPI */
  MPI_Finalize();
	
  return 0;
}

File: mpi/exercise/pi_mpi.f90
program pi_mpi

  use mpi

  implicit none
  integer, parameter :: dp=selected_real_kind(14)
  integer :: i, ierr, nproc, myrank, istart, iend
  integer, parameter :: n=100000000
  real(dp) :: x,pi,sum,total_sum, step,start_time,end_time
  
  
  ! MPI Calls
  ! Initialize MPI
  call MPI_INIT(ierr)
  ! Get Process ID
  call MPI_COMM_RANK(MPI_COMM_WORLD,myrank,ierr)
  ! Get Process number
  call MPI_COMM_SIZE(MPI_COMM_WORLD,nproc,ierr)

  sum = 0d0
  step = 1.d0/real(n,kind=dp)

  ! Start timer only on main
  if (myrank == 0 ) then
     start_time = MPI_WTIME();
  end if
  
  ! Broadcast value of n to all processes, not sure why this gives a seg fault
  !call MPI_BCAST(n,1,MPI_INTEGER,0,MPI_COMM_WORLD,ierr)
  
  ! Divide work such that each process gets n/n_p elements to calculate
  do i = myrank+1, n, nproc
     x = (real(i) + 0.5d0) * step
     sum = sum + 4.d0 / (1.d0 + x ** 2)
  end do

  ! collect all partial results with a reduction operation
  

  ! Synchronize all processes
  

  ! Calculate value of pi and stop timer if main process
  if (myrank == 0) then
     end_time = MPI_WTIME();
     pi = step * total_sum
     print '(a,f17.15)', "pi = ", pi
     print '(a,f9.3,a)', "time to compute =",end_time - start_time, " seconds"
  end if
  
  ! Finalize MPI
  call MPI_FINALIZE(ierr)
  
end program pi_mpi
